WIP - Extended reports

danielb987 opened the pull request.

danielb987 commented:
@pabender 
What is `jmri.jmrix.internal.TrackReporter` used for? If it's used? I'm aware that `ProxyReporterManager` creates a `TrackReporter` if the system name doesn't match a connection. But does `TrackReporter` serves any other purpose except being the default reporter?

I ask the question since I'm trying to refactorize the reporting infrastructure in JMRI and go from using `Object` to `jmri.ExtendedReport`. And `TrackReporter` has public methods `pushEast()`, `pushWest()`, `pullEast()` and `pullWest()` which takes an Object as parameter. These methods doesn't seem to be used by anyone except `TrackReporterTest`.

danielb987 commented:
@bobjacobsen 
Can you look at `RpsReporter` and `RpsReport` in this PR? Does my changes to `RpsReporter` makes sense?

For `RpsReport`, I have somewhat redefined the physical location. If the user has set a physical location for the reporter bean, that physical location is used. Otherwise, the x, y, z coordinates of the reporter are used.

I have removed several methods in `RpsReporter` that aren't used and moved them to `RpsReport`. Is this okey or may it cause trouble to users?

bobjacobsen commented:
The RPS stuff looks OK to me.  AFAIK, there are only a couple users of that hardware still (it hasn't been made for quite a while, and there only ever were ~20 sets sold). If they encounter issues, we can fix it then.

pabender commented:
> @pabender What is `jmri.jmrix.internal.TrackReporter` used for? If it's used? I'm aware that `ProxyReporterManager` creates a `TrackReporter` if the system name doesn't match a connection. But does `TrackReporter` serves any other purpose except being the default reporter?
> 
> I ask the question since I'm trying to refactorize the reporting infrastructure in JMRI and go from using `Object` to `jmri.ExtendedReport`. And `TrackReporter` has public methods `pushEast()`, `pushWest()`, `pullEast()` and `pullWest()` which takes an Object as parameter. These methods doesn't seem to be used by anyone except `TrackReporterTest`.

They are there to allow keeping track of what is in a block in order, and are primarily there for use in scripts.

I have a script that uses them that isn’t a part of the JMRI distribution:
https://github.com/pabender/Sensys2014/blob/master/ReporterTrackContents.py

The layout this script was written has a single RFID reader at the entrance to a yard and switch positions are used to determine which track we are accessing.  The script pushes or pulls as is appropriate.

The reason there are east and west push/pull methods is because if you are doing this on a double ended siding, you can add or remove cars from either end.

Incidentally, this isn’t part of the CollectingReporter interface because the collectingReporter interface doesn’t dictate what type of collection is used ( the other implementations are for LocoNet, which uses a set, and The Roco Z21 CAN reporters, which use an ArrayList ).  The collections used are dictated by the behavior of the underlying hardware.

danielb987 commented:
@pabender 
I have restored the methods in `TrackReporter`, but I have changed the report from `Object` to `jmri.ExtendedReport`. That makes the collection checked. To aid scripts like yours, I have added the class `ExtendedReport.StringReport`. This means you need to change:
`tr.pushWest("Hello");`
to
`tr.pushWest(new ExtendedReport.StringReport("Hello"));`

and

`Assert.assertEquals("after 1st push","Hello",tr.getCurrentReport());`
to
`Assert.assertEquals("after 1st push","Hello",tr.getCurrentReport().toString());`

I understand that this isn't ideal for script writers, but I try to balance between having a much better reporting infrastructure in JMRI and to make the transition relative easy for script writers.

pabender commented:
> @pabender I have restored the methods in `TrackReporter`, but I have changed the report from `Object` to `jmri.ExtendedReport`. That makes the collection checked. To aid scripts like yours, I have added the class `ExtendedReport.StringReport`. This means you need to change: `tr.pushWest("Hello");` to `tr.pushWest(new ExtendedReport.StringReport("Hello"));`

Actually, the script i pointed you to works with the reported objects.   They are only called inputString in one method due to historical reasons.

> `Assert.assertEquals("after 1st push","Hello",tr.getCurrentReport());` to `Assert.assertEquals("after 1st push","Hello",tr.getCurrentReport().toString());`
> 
> I understand that this isn't ideal for script writers, but I try to balance between having a much better reporting infrastructure in JMRI and to make the transition relative easy for script writers.

I understand that problem, though in the test case you note above, there isn’t anything magical about pushing a string and checking the value.  It can be an extendedReport object.  It was only a string because the interface stated objects were allowed.

danielb987 commented:
> Actually, the script i pointed you to works with the reported objects. They are only called inputString in one method due to historical reasons.

I have restored `TrackReporter` to how it was before this PR. The issue was that I have changed `AbstractReporter.setReport()` to throw an UnsupportedOperationException if the method `isExtendedReportsSupported()` returns true. But I have now changed `TrackReporter` so that `TrackReporter.isExtendedReportsSupported()` returns false. And therefore these changes are not needed.

Thanks for your comments on this.

danielb987 commented:
> > @pabender What is `jmri.jmrix.internal.TrackReporter` used for? If it's used? I'm aware that `ProxyReporterManager` creates a `TrackReporter` if the system name doesn't match a connection. But does `TrackReporter` serves any other purpose except being the default reporter?
> > I ask the question since I'm trying to refactorize the reporting infrastructure in JMRI and go from using `Object` to `jmri.ExtendedReport`. And `TrackReporter` has public methods `pushEast()`, `pushWest()`, `pullEast()` and `pullWest()` which takes an Object as parameter. These methods doesn't seem to be used by anyone except `TrackReporterTest`.
> 
> They are there to allow keeping track of what is in a block in order, and are primarily there for use in scripts.
> 
> I have a script that uses them that isn’t a part of the JMRI distribution: https://github.com/pabender/Sensys2014/blob/master/ReporterTrackContents.py
> 
> The layout this script was written has a single RFID reader at the entrance to a yard and

As I said earlier today, I have restored `TrackReporter` as it was before. But I would like to do a small refactorization that will need a small change in your script:

Today, there is no guarantee for the user that `InstanceManager.getDefault(ReporterManager).provideReporter(name)` actually returns a `TrackReporter`. Therefore, any developer could easily remove or change `TrackReporter` without thinking that it would affect users. Especially sine `TrackReporter` is in the package `jmri.jmrix.internal`. I don't think any user should rely on `jmri.jmrix.internal`.

I would like to:
- Add the inner interface `TrackReporter` to `CollectingReporter` and let it extend `CollectingReporter`. It will have the requirement that it's collection is ordered. The new interface `TrackReporter` would have the methods:
```
public void pushEast(Object o);
public void pushWest(Object o);
public Object pullEast();
public Object pullWest();
```

- Add the method `ReporterManager.provideTrackReporter(@Nonnull String name)` that provides a `jmri.CollectingReporter.TrackReporter`

And then either
- Let `InternalReporterManager.provideReporter(@Nonnull String name)` provide a minimalistic internal reporter that doesn't do anything

or

- Let `InternalReporterManager.provideReporter(@Nonnull String name)` throw an exception like `UnsupportedOperationException` if that's allowed.

This means that your script would need to call
`jmri.InstanceManager.getDefault(jmri.ReporterManager).provideTrackReporter(reporterName).addPropertyChangeListener(self)`
instead of
`jmri.InstanceManager.getDefault(jmri.ReporterManager).provideReporter(reporterName).addPropertyChangeListener(self)`

pabender commented:
> > > @pabender What is `jmri.jmrix.internal.TrackReporter` used for? If it's used? I'm aware that `ProxyReporterManager` creates a `TrackReporter` if the system name doesn't match a connection. But does `TrackReporter` serves any other purpose except being the default reporter?
> > > I ask the question since I'm trying to refactorize the reporting infrastructure in JMRI and go from using `Object` to `jmri.ExtendedReport`. And `TrackReporter` has public methods `pushEast()`, `pushWest()`, `pullEast()` and `pullWest()` which takes an Object as parameter. These methods doesn't seem to be used by anyone except `TrackReporterTest`.
> > 
> > They are there to allow keeping track of what is in a block in order, and are primarily there for use in scripts.
> > I have a script that uses them that isn’t a part of the JMRI distribution: https://github.com/pabender/Sensys2014/blob/master/ReporterTrackContents.py
> > The layout this script was written has a single RFID reader at the entrance to a yard and
> 
> As I said earlier today, I have restored `TrackReporter` as it was before. But I would like to do a small refactorization that will need a small change in your script:
> 
> Today, there is no guarantee for the user that `InstanceManager.getDefault(ReporterManager).provideReporter(name)` actually returns a `TrackReporter`. Therefore, any developer could easily remove or change `TrackReporter` without thinking that it would affect users. Especially sine `TrackReporter` is in the package `jmri.jmrix.internal`. I don't think any user should rely on `jmri.jmrix.internal`.
> 

I think that flies in the face of the way things have traditionally worked, i.e. users have always relied on internal for various things,  and sometimes internal supports operations that real hardware doesn't.

> I would like to:
> 
> * Add the inner interface `TrackReporter` to `CollectingReporter` and let it extend `CollectingReporter`. It will have the requirement that it's collection is ordered. The new interface `TrackReporter` would have the methods:
> 
> ```
> public void pushEast(Object o);
> public void pushWest(Object o);
> public Object pullEast();
> public Object pullWest();
> ```

Making an interface is fine, and If I were writing this today, I would have made one.   I would NOT make the TrackReporter interface extended CollectingReporter however.  In other words, It SHOULD be possible for a reporter to provide the ordered reporting without exposing the underlying collection the reporter is using.

> * Add the method `ReporterManager.provideTrackReporter(@Nonnull String name)` that provides a `jmri.CollectingReporter.TrackReporter`
> 

I find this part of your plan highly problematic.  Client code should be able to always call ReporterManager.providerReporter(name) and get whatever reporter is provided by the ReporterManager instance.

The client code can then check to see if it implements the interfaces required.

i.e. if the new interface is jmri.TrackReporter, this would allow the client code to do something like this:

reporter = InstanceManager.getDefault(ReporterManager.class).provideReporter("foo").
if(reporter instanceof jmri.CollectingReporter){
   // do something with the collection
}
if(reporter instanceof jmri.TrackReporter){
   // push and pull as required
}

There could be a TrackReporterDecorator provided.  The decorator would implement the jmri.Reporter and jmri.TrackReporter interfaces.  It would handle the TrackReporter details itself, but the Reporter details would be passed on to the underlying reporter implementation.

and the client could do something like this:

reporter = new jmri.TrackReporter(InstanceManager.getDefault(ReporterManager.class).provideReporter("foo")).
   // push and pull as required

(that construction could still occur inside the internal manager to provide the same behavior we have today, but it
would split the base operations of the internal reporter and the TrackReporter to maintain single responsibility).

The whole idea here is to make the code better adhere to the SOLID principles, which it doesn't right now.

danielb987 commented:
@pabender 
Thanks for your help. I have read your reply and decided to not refactor `TrackReporter`, at least not now.

But maybe it would be good to remove `CollectingReporter` from `TrackReporter` now? That removes the method `getCollection()` so it doesn't expose the underlying collection the reporter is using. Do you think I should do this?

----

`reporter = new jmri.TrackReporter(InstanceManager.getDefault(ReporterManager.class).provideReporter("foo")).`

If I'm going to implement this, I would need `jmri.TrackReporter` to extend `jmri.implementation.AbstractReporter`. And it's not good to have classes in the `jmri`package to depend on the `jmri.implementation` package.

pabender commented:
> @pabender Thanks for your help. I have read your reply and decided to not refactor `TrackReporter`, at least not now.
> 
> But maybe it would be good to remove `CollectingReporter` from `TrackReporter` now? That removes the method `getCollection()` so it doesn't expose the underlying collection the reporter is using. Do you think I should do this?

I wouldn’t now, because the display code for reporter’s doesn’t handle the collections, so if you want to find the full contents of the track, you currently need access to the collection.

> `reporter = new jmri.TrackReporter(InstanceManager.getDefault(ReporterManager.class).provideReporter("foo")).`
> 
> If I'm going to implement this, I would need `jmri.TrackReporter` to extend `jmri.implementation.AbstractReporter`. And it's not good to have classes in the `jmri`package to depend on the `jmri.implementation` package.

I have it in the wrong spot above.  It should be jmri.implementation.decorators.TrackReporter….  But the decorator should not extend AbstractReporter.  The jmri.implementation.decorators.TimeoutReporter provides an example of how this works.

danielb987 commented:
> I wouldn’t now, because the display code for reporter’s doesn’t handle the collections, so if you want to find the full contents of the track, you currently need access to the collection.

I have removed the dependency of `CollectingReporter` and instead added the method `getTrackContent()` that returns an unmodifiable collection of the track contents.

pabender commented:
> > I wouldn’t now, because the display code for reporter’s doesn’t handle the collections, so if you want to find the full contents of the track, you currently need access to the collection.
> 
> I have removed the dependency of `CollectingReporter` and instead added the method `getTrackContent()` that returns an unmodifiable collection of the track contents.

Wait a second… you need to revert that change.  CollectingReporter is not a dependency.  It is an interface, and that interface correctly describes the functionality the track reporter provides 

You added a method that isn’t part of an interface that does the same job the one method defined in the CollectingReporter interface is intended to do.  This means that the TrackReporter as you have modified it has the same functionality as classes that implement the CollectingReporter interface. But TrackReporter is no longer interchangeable with other classes that also implement the CollectingReporter interface for purposes of getting the full Collection and it should be interchangeable with those classes ( this applies to the LocoNet Reporters and the Roco Z21 Can Reporters. )

I do think it would be ok to say the One method defined in the CollectingReporter interface returns a copy of the underlying collection.

Your  change is also a breaking change for external code and scripts that work with the collections ( such as https://github.com/JMRI/JMRI/blob/14c53f7cbbc2769b97fa8388f6142aa555a21574/jython/CollectingReporterExample.py )

pabender reviewed (changes requested):
You’ll need to make the TracKReporter handle the extended reports and restore its implementation of CollectingReporter.

pabender commented:
@danielb987 i requested changes, because I think you are breaking the usecase the TrackReporter was designed to handle, namely how do report what is on the track when there is no physical reporter on that track.

In particular, the scripts I pointed to earlier in this thread use the TrackReporter in a yard where there is one physical reporter at the entrance to the yard and the positions of the turnouts in the yard are known… the actual layout this is used on is below:

![](https://nscaleintermodal.com/layout/RFIDTestLayout/Photos/IMGP6578.JPG)

The TrackReporter needs to keep track of whatever the physical reporter keeps track of, and it should interact with the physical reporters the same way any other reporter does ( some of the code you removed breaks one interaction which is that if reporter A sees a tag and then reporter B sees the same tag, it is to be removed from reporter A).

I don’t care what the underlying object is representing the tag ( I.e. this is technology neutral ).  What I care about is that on the layout above ( or any similar configuration ) If the physical reporter shows car ABC1234 and the switches are lined for track 2, then car ABC1234 shows up on the TrackReporter for track 2.

The fact that this has only been done in scripts now is somewyour LogixNG should be able to handle this usecase

danielb987 commented:
@pabender 
I will revert TrackReporter.
